### PR TITLE
documentation reorganization

### PR DESCRIPTION
**Pull request by maccum**
based on #4061, and part of a broader move toward the four types of documentation 

1. Tutorials
2. Reference/API 
3. Explanation
4. How-To Guides

This just moves some stuff around between sections 1 and 3. I'll put the beginning of the How-To Guides in a separate PR. 

- change layout of tutorials landing page so that clicking on the tutorial title in the sidebar brings you to the actual tutorial
- delete Types tutorial and Missingness tutorial, which weren't really tutorials, and move some of their information to the Types and Missingness sections of Section III.
- break the "Overview" page into smaller subpages, and rename it to Hailpedia. This is section III of the documentation, but I don't like 'Explanation' or 'Overview' as a title. Other suggestions welcome.
- a few other edits to add additional explantation to some of the jupyter tutorials

### CODE REVIEW

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+It is possible to represent matrix data by coordinate in a table , storing one
+record per entry of the matrix. However, the :class:`.MatrixTable` represents
+this data far more efficiently and exposes natural interfaces for computing on
+it.
+
+The :meth:`.MatrixTable.rows` and :meth:`.MatrixTable.cols` methods return the
+row and column fields as separate tables. The :meth:`.MatrixTable.entries`
+method returns the matrix as a table in coordinate form -- use this object with
+caution.
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: ... because this representation is costly to compute and is significantly larger in memory.

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+Keys
+====
+
+Matrix tables have keys just as tables do. However, instead of one key, matrix
+tables have two keys: a row key and a column key. Row fields are indexed by the
+row key, column fields are indexed by the column key, and entry fields are
+indexed by the row key and the column key. The key structs can be accessed with
+:attr:`.MatrixTable.row_key` and :attr:`.MatrixTable.col_key`. It is possible to
+change the key with :meth:`.MatrixTable.key_rows_by` and
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: change the *keys*

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+:meth:`.MatrixTable.key_cols_by`.
+
+Note that changing the row key, however, may be an expensive operation.
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: not sure the however is necessary. I bias towards giving folks justifications for our statements. I would say
> Due to the data representation of a matrix table, changing a row key is often an expensive operation.

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+Hail matrix tables are natively distributed objects, and as such have another
+key: a partition key. This key is used for specifying the ordering of the matrix
+table along the row dimension, which is important for performance. Access this
+with :attr:`.MatrixTable.partition_key`
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: This is probably going away soon enough that we should just elide it. @patrick-schultz thoughts?

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+==================
+
+All fields (row, column, global, entry) are top-level and exposed as attributes
+on the :class:`.MatrixTable` object. For example, if the matrix table `mt` had a
+row field `locus`, this field could be referenced with either ``mt.locus`` or
+``mt['locus']``. The former access pattern does not work with field names with
+spaces or punctuation.
+
+The result of referencing a field from a matrix table is an :class:`.Expression`
+which knows its type and knows its source as well as whether it is a row field,
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: "source" is used without definition here, but maybe that's ok... not sure, maybe "its source matrix table or table"

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+When evaluated in a Python interpreter, we can see ``mt.locus`` is a
+:class:`.LocusExpression` with type ``locus<GRCh37>`` and it is a row field of
+the MatrixTable `mt`.
+
+    >>> mt
+    <hail.matrixtable.MatrixTable at 0x1107e54a8>
+
+    >>> mt.locus
+    <LocusExpression of type locus<GRCh37>>
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: The evaluation does not allow us to see that it is a row field of the mt, I'd leave that bit off.

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+Likewise, ``mt.DP`` would be an :class:`.Int32Expression` with type ``int32``
+and is an entry field of `mt`. It is indexed by both rows and columns as denoted
+by its indices when describing the expression:
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: "describe" is used off-hand here, but I think it's worth introducing it properly. Maybe:

> Hail expressions can also `describe` themselves, providing information on the source matrix table or table and which keys index the expression, if any. For example, ``mt.DP.describe()`` tells us that ``mt.DP`` has type ``int32`` and is an entry field of ``mt``:

(also mt should be double back-ticked)

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+        'locus': locus<GRCh37>
+        'alleles': array<str>
+    Partition key:
+        'locus': locus<GRCh37>
+    ----------------------------------------
+
+Common Operations
+=================
+
+Like tables, Hail provides a number of useful methods for manipulating data in a
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: I tend to avoid words like "useful" since they don't actually convey any, well, useful information :P

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+mean of the `GQ` field, and removes columns where the result is smaller than 20.
+
+**Annotate**
+
+:class:`.MatrixTable` has four methods to add new fields or update existing fields:
+
+- :meth:`.MatrixTable.annotate_rows`
+- :meth:`.MatrixTable.annotate_cols`
+- :meth:`.MatrixTable.annotate_entries`
+- :meth:`.MatrixTable.annotate_globals`
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: kind of think globals should be first?

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+of non-missing entries `GT` per row:
+
+    >>> mt_new = mt.annotate_rows(call_rate = hl.agg.fraction(hl.is_defined(mt.GT)))
+
+Annotate methods are also useful for updating values. For example, to update the
+GT entry field to be missing if `GQ` is less than 20, we can do the following:
+
+    >>> mt_new = mt.annotate_entries(GT = hl.case()
+    ...                                     .when(mt.GQ >= 20, mt.GT)
+    ...                                     .or_missing())
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: we should probably take this chance to encourage `hl.or_missing(mt.GQ >= 20, mt.GT)`

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+    struct {
+        GT: call
+    }
+
+:class:`.MatrixTable` has four select methods that select and create new fields:
+
+- :meth:`.MatrixTable.select_rows`
+- :meth:`.MatrixTable.select_cols`
+- :meth:`.MatrixTable.select_entries`
+- :meth:`.MatrixTable.select_globals`
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: again, I sort of like globals first, but 🤷‍♀️

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+    >>> mt_new = mt.annotate_entries(GT = hl.case()
+    ...                                     .when(mt.GQ >= 20, mt.GT)
+    ...                                     .or_missing())
+
+**Select**
+
+Select is used to create a new schema for a dimension of the matrix table. For
+example, following the matrix table schemas from importing a VCF file (shown above),
+to create a hard calls dataset where each entry only contains the `GT` field
+one can do the following:
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: I think this paragraph should mention that keys are always preserved.

Comment (danking) [reply python/hail/docs/hailpedia/matrixtable.rst]: or maybe it should be more loud below?

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+- :meth:`.MatrixTable.select_cols`
+- :meth:`.MatrixTable.select_entries`
+- :meth:`.MatrixTable.select_globals`
+
+Each method can take either strings referring to top-level fields, an attribute
+reference (useful for accessing nested fields), as well as keyword arguments
+``KEY=VALUE`` to compute new fields. The Python unpack operator ``**`` can be
+used to specify that all fields of a Struct should become top level fields.
+However, be aware that all top-level field names must be unique. In this
+example, `**mt['info']` would fail because `DP` already exists as an entry
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: this example => the next example
?

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+field.
+
+The example below will keep the row keys `locus` and `alleles` as well as add
+two new fields: `AC` is making the subfield `AC` into a top level field and
+`n_filters` is a new computed field.
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: this sentence reads weird. maybe: 

> as well as add two new fields. ``AC = mt.info.AC`` is making the subfield ``AC`` into a top level field. The second keyword argument is computing the new field `n_filters`.

eh. not sure any further explanation after "add two new fields" is necessary.

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+These methods take an aggregated expression and evaluate it, returning
+a Python value.
+
+An example of querying entries is to compute the global mean of field `GQ`:
+
+    >>> mt.aggregate_entries(hl.agg.mean(mt.GQ))
+    67.73196915777027
+
+It is possible to compute multiple values simultaneously (and encouraged,
+because grouping two computations together will run twice as fast!) by
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: running twice as fast is not strictly true, maybe we can say something like "because grouping two computations together is far more efficient, traversing the dataset only once rather than twice.".

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+            mean: float64,
+            stdev: float64,
+            min: float64,
+            max: float64,
+            n: int64,
+            sum: float64
+        }
+    }
+
+    >>> print(mt_grouped.col.dtype)
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: I kind of think `mt_grouped.describe()` is the most informative thing here.

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+- :meth:`.MatrixTable.union_cols`
+- :meth:`.MatrixTable.union_rows`
+
+:meth:`.MatrixTable.union_cols` joins matrix tables together by performing an
+inner join on rows while concatenating columns together (similar to `paste` in
+Unix). Likewise, :meth:`.MatrixTable.union_rows` performs an inner join on
+columns while concatenating rows together (similar to `cat` in Unix).
+
+In addition, Hail provides support for joining data from multiple sources together
+if the keys of each source are compatible (same order and type, but the names do
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: "order" is very unclear here, the fact that a "key" is a composition of several fields is not really clear from any of the prior verbiage. Hmm. It's hard to describe this. maybe: "compatible (meaning that, viewing the key as an ordered tuple of fields, those fields must have the same type, but need not have the same field names)"

File: python/hail/docs/hailpedia/matrixtable.rst [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+not need to be identical) using Python's bracket notation ``[]``. The arguments
+inside the brackets are the destination key as a single value or a tuple if there
+are multiple destination keys.
+
+For example, we can annotate rows with row fields from another matrix table or
+table. Let `gnomad_data` be a :class:`.Table` keyed by two row fields with type
+``locus`` and ``array<str>``, which matches the row keys of `mt`:
+
+    >>> mt_new = mt.annotate_rows(gnomad_ann = gnomad_data[mt.locus, mt.alleles])
```
Comment (danking) [inline python/hail/docs/hailpedia/matrixtable.rst]: I think I'd start with the example, the paragraph before last is a bit hard to parse.

Comment (danking) [reply python/hail/docs/hailpedia/matrixtable.rst]: thought the previous paragraph works well. Maybe its just the one before last that's hard to parse

**Comment by maccum** [comment]
I think I'm going to split this PR up into a few separate PRs, so that it will be easier to review.

File: python/hail/docs/hailpedia/matrixtable.rst (hunk 4 above)
Comment (patrick-schultz) [reply python/hail/docs/hailpedia/matrixtable.rst]: I agree.

File: python/hail/docs/hailpedia/matrixtable.rst (hunk 16 above)
Comment (maccum) [reply python/hail/docs/hailpedia/matrixtable.rst]: The schema of mt_grouped is really large. I find it a bit distracting to show the whole output of ``describe``.
